src/wallet/wallet.cpp

```diff
@@ -2105,10 +2108,17 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, int nConfMine, int
     return true;
 }
 
-bool CWallet::SelectCoins(const CAmount& nTargetValue, set<pair<const CWalletTx*,unsigned int> >& setCoinsRet, CAmount& nValueRet, const CCoinControl* coinControl) const
+bool CWallet::SelectCoins(const CAmount& nTargetValue, set<pair<const CWalletTx*,unsigned int> >& setCoinsRet, CAmount& nValueRet,  bool& fOnlyCoinbaseCoinsRet, const CCoinControl* coinControl) const
 {
     vector<COutput> vCoins;
-    AvailableCoins(vCoins, true, coinControl);
+    AvailableCoins(vCoins, true, coinControl, false, false); // we don't want coinbase coins
+    if (vCoins.size()>0) {
+        fOnlyCoinbaseCoinsRet = false;
+    } else {
+        vector<COutput> vAllCoins;
+        AvailableCoins(vAllCoins, true, coinControl, false, true);
+        fOnlyCoinbaseCoinsRet = vAllCoins.size()>0;
+    }
```

Comment:
If I read the code correctly it has this behavior:

- If all my coins are coinbases, I'll get the error message "Coinbase funds can only be sent to a zaddr".
- Otherwise, (when I have some coinbases and some non-coinbases, but the non-coinbases are insufficient) I'll get the error "Insufficient funds."

This means that in the case where my non-coinbases aren't enough, but my non-coinbases *plus* my coinbases would be, then I still get the error "Insufficient funds." Should that be a separate case, i.e. "You have sufficient funds, but coinbases must be spent to a zaddr first"?

There might be TOCTTOU bug: the wallet lock is acquired inside `AvailableCoins()` so between the first call and the second, and the wallet state might change in between. `fOnlyCoinbaseCoinsRet` would be set incorrectly if new non-coinbase coins were acquired. I'm not sure if that matters.